File: website/backend/app/models/catalog.py
from __future__ import annotations

from typing import Literal

from pydantic import Field

from backend.app.models.shared import EventEnvelope, StrictModel


CatalogLoadStatus = Literal["idle", "loading", "ready", "partial", "unavailable"]
CatalogItemStatus = Literal["active", "hidden", "unavailable"]
AvailabilityState = Literal["available", "unavailable"]
GarmentSilhouette = Literal["upper-body"]


class GarmentCategory(StrictModel):
    id: str
    label: str
    sortOrder: int


class GarmentSizeOption(StrictModel):
    code: str
    label: str
    availability: AvailabilityState


class GarmentColorOption(StrictModel):
    id: str
    label: str
    variantId: str
    availability: AvailabilityState
    swatchHex: str | None = None


class Garment(StrictModel):
    id: str
    sku: str
    name: str
    categoryId: str
    silhouette: GarmentSilhouette
    status: CatalogItemStatus
    description: str | None = None
    heroImageUrl: str | None = None
    availableSizes: list[GarmentSizeOption]
    availableColors: list[GarmentColorOption]
    defaultVariantId: str | None = None


class CatalogSnapshotPayload(StrictModel):
    status: CatalogLoadStatus
    categories: list[GarmentCategory]
    garments: list[Garment]


class CatalogSnapshotUpdatedEvent(EventEnvelope):
    type: Literal["catalog.snapshot.updated"]
    source: Literal["catalog"] = "catalog"
    payload: CatalogSnapshotPayload


class CatalogSnapshotUnavailableEvent(EventEnvelope):
    type: Literal["catalog.snapshot.unavailable"]
    source: Literal["catalog"] = "catalog"
    payload: dict[Literal["status"], Literal["unavailable"]] = Field(
        default_factory=lambda: {"status": "unavailable"}
    )

File: website/src/types/catalog.ts
import type {
  AvailabilityState,
  CategoryId,
  ColorId,
  GarmentId,
  GarmentVariantId,
  IsoTimestamp,
  SizeCode,
} from '@/types/shared';

export type CatalogLoadStatus =
  | 'idle'
  | 'loading'
  | 'ready'
  | 'partial'
  | 'unavailable';
export type CatalogItemStatus = 'active' | 'hidden' | 'unavailable';
export type GarmentSilhouette = 'upper-body';

export type GarmentCategory = {
  id: CategoryId;
  label: string;
  sortOrder: number;
};

export type GarmentColorOption = {
  id: ColorId;
  label: string;
  swatchHex?: string;
  variantId: GarmentVariantId;
  availability: AvailabilityState;
};

export type GarmentSizeOption = {
  code: SizeCode;
  label: string;
  availability: AvailabilityState;
};

export type Garment = {
  id: GarmentId;
  sku: string;
  name: string;
  categoryId: CategoryId;
  silhouette: GarmentSilhouette;
  status: CatalogItemStatus;
  description?: string;
  heroImageUrl?: string;
  availableSizes: GarmentSizeOption[];
  availableColors: GarmentColorOption[];
  defaultVariantId?: GarmentVariantId;
};

export type GarmentSelection = {
  garmentId: GarmentId;
  variantId?: GarmentVariantId;
  sizeCode?: SizeCode;
  colorId?: ColorId;
  selectedAt: IsoTimestamp;
};

export type CatalogFocus = {
  categoryId?: CategoryId;
  highlightedGarmentId?: GarmentId;
};
